### PR TITLE
Monkeypatch pytest xml file race condition

### PR DESCRIPTION
**Pull request by Kadino**
Signed-off-by: sweeneys <sweeneys@amazon.com>

## What does this PR do?

Attempts a new workaround to resolve a filesystem race condition in pytest https://github.com/o3de/o3de/issues/13720 . The race condition most frequently appears when running multiple instances of pytest in parallel, or when invoking pytest inside another instance of pytest.

The change copies existing pytest code from https://github.com/pytest-dev/pytest/blob/7.2.x/src/_pytest/junitxml.py#L646 and modifies a single line to include `exist_ok=True`. This notably does not work around another edge case where the user requests creating a file at `grandparent/parent/file.xml` but a non-directory file already exists at `grandparent/parent` . That case does not appear to be reproducing in O3DE, should be very uncommon, and would likely require renaming the user's requested file path.

This fix was hooked into the pytest plugin for configuration because it cannot be done in an autouse pytest fixture. Unfortunately all fixtures are instantiated after pytest initializes a LogXML object.

A better fix would be to directly fix pytest itself. After proving stability, this fix will be recommended to the pytest maintainers.

## How was this PR tested?

Executed the unit tests of LyTestTools with the --junitxml flag set, to make sure the patched method is being called. Manually set a breakpoint and manually recreated the race condition of the file structure changing between `os.path.isdir` and `os.makedirs`.

Will use Jenkins runs to collect additional data on whether we continue to encounter race conditions here.

### CODE REVIEW

**Review by evanchia-ly-sdets — APPROVED** [review]
LGTM, but I was wondering how we could prevent this from breaking if pytest ever updates their function?

**Comment by Kadino** [comment]
> I was wondering how we could prevent this from breaking if pytest ever updates their function?

We'd have to co-maintain this monkeypatch when we upgrade pytest versions. At least until we get the pytest maintainers to merge a fix. I want a little more data to prove it works before submitting them a bug and/or PR.

File: Tools/LyTestTools/ly_test_tools/_internal/pytest_plugin/test_tools_fixtures.py
Comment (Kadino) [inline Tools/LyTestTools/ly_test_tools/_internal/pytest_plugin/test_tools_fixtures.py:58]: Oops, left myself a TODO in which I intended to link to a GHI in the pytest project.

**Comment by evanchia-ly-sdets** [comment]
> > I was wondering how we could prevent this from breaking if pytest ever updates their function?
> 
> We'd have to co-maintain this monkeypatch when we upgrade pytest versions. At least until we get the pytest maintainers to merge a fix _**and**_ we update to consume their fix. I want a little more data to prove it works before submitting them a bug and/or PR.

Could we perhaps comment which pytest version we're on to make sure we're up to date on co-maintaining or something of that effect?

**Comment by Kadino** [comment]
> Could we perhaps comment which pytest version we're on to make sure we're up to date on co-maintaining or something of that effect?
Line 431

**Comment by Kadino** [comment]
Appears to have created an issue with cached python files. Testing a different type of fix, but this monkeypatch approach would not be suitable if these problems persist.

**Comment by Kadino** [comment]
Clean AR build proved second monkeypatch approach works. Have not found additional race conditions in local testing.

**Comment by Kadino** [comment]
Additional local testing reproduced no other race conditions, so I proposed a PR to the pytest maintainers. This monkeypatch can be removed after that code gets packaged, released, and o3de updates its requirements.txt

**Review by AMZN-scspaldi — APPROVED** [review]

**Pull request merged**
